### PR TITLE
executor: avoid sum from avg overflow (#30010)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #30010 to release-5.2
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/30378
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/30378:release-5.2-9aa756336e7b
```

---

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/29952

Problem Summary: 
1. the sum from avg may overflow, we should infer the type as the normal sum.
2. the previous PR https://github.com/pingcap/tidb/pull/20017 is a hack to fix the [issue](https://github.com/pingcap/tidb/issues/19426) , we should let the final agg output the target return type which is inferred in TypeInfe() for Agg.
3. typeInfer4xxx should valid the precision and scale of decimal type.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: executor/aggfuncs/builder.go [hunk 1]
```diff
@@ -17,9 +17,12 @@ import (
 	"fmt"
 	"strconv"
 
+<<<<<<< HEAD
 	"github.com/cznic/mathutil"
 	"github.com/pingcap/parser/ast"
 	"github.com/pingcap/parser/mysql"
+=======
+>>>>>>> 9aa756336... executor: avoid sum from avg overflow (#30010)
```
Comment (hawkingrei) [inline executor/aggfuncs/builder.go:25]: need to remove it.

Comment (fzhedu) [reply executor/aggfuncs/builder.go]: done

**Review by hawkingrei — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Review by LittleFall — APPROVED** [review]

**Pull request merged**